Adds Broadcast model

aaronschachter opened the pull request:
#### What's this PR do?
* Adds a `broadcasts` collection to associate a Mobile Commons broadcast with its corresponding DS Campaign, broadcast type, and `messages` object
    * Currently only supports a `signup_prompt_declined` message, to display via `mobilecommons.send_message` when User responds with a NO to the Signup broadcast. 
    * Deprecates the CampaignBot `msg_signup_broadcast_declined` message

* Refactors Signup broadcast to load its Campaign by finding a record for the `broadcast_id` in the `broadcasts` table, deprecating the `CAMPAIGNBOT_BROADCAST_CAMPAIGN` config variable (#718)
    * This allows to send multiple Signup broadcasts throughout the week without needing to worry about setting the correct config variable

* Does *NOT* add support for Reminder broadcasts (where the yes/no response should start/continue reportback draft), just gets refactor of Signups working in anticipation of Reminder support

* Adds logging, `MOBILECOMMONS_DISABLED` check to `mobilecommons.send_message` function introduced in #763 

#### How should this be reviewed?
* Add a document into `broadcasts` for a broadcast id, and a CampaignBot campaign X (`type` doesn't really matter because we're not using it yet)
* Post against `chatbot` endpoint passing the `broadcast_id` you added a record for. Confirm a YES response will create the Signup for Campaign X, NO responses returns Campaignbot `signup_broadcast_declined`



#### Any background context you want to provide?
* Reminder broadcasts functionality still left to do #742.

#### Relevant tickets
#742 

#### Checklist
- [x] Documentation added for new features/changed endpoints.
- [x] Tested on staging.
- [x] Manually created test`broadcasts` document in staging DB
- [x] Manually created test `broadcasts` document in production DB

aaronschachter committed:
Adds Broadcast model

aaronschachter commented on app/routes/chatbot.js:
```diff
@@ -112,74 +113,82 @@ router.post('/', (req, res) => {
+              campaign = app.locals.campaigns[broadcast.campaign];
+
+              if (!campaign) {
+                const err = new Error('broadcast campaign undefined');
```
TODO: Trigger stathat (with email alert) -- broadcast / mData won't work if we don't have a campaign to signup or reportback for.

aaronschachter commented:
If this branch gets merged we'll also need to update the https://github.com/DoSomething/gambit/wiki/Mobile-Commons#signups documentation to deprecate use of the `CAMPAIGNBOT_BROADCAST_CAMPAIGN` variable and add steps for implementation:
1. Create the desired Mobile Commons Broadcast in Mobile Commons, generating a Broadcast ID
2. Use that ID as the `_id` in the corresponding `broadcasts` document in the Gambit DB, adding the `campaign` we want to Signup for
3. Test by posting to `v1/chatbot`, passing the `broadcast_id` -- verifying the Signup is created for desired Campaign

aaronschachter committed:
Adds debug statements to print broadcast, campaign id's

aaronschachter committed:
Handle broadcast not found

aaronschachter committed:
Adds signup_prompt_declined message to Broadcast

aaronschachter committed:
Adds logging, check for disabled flag

aaronschachter commented:
To test this branch, the `broadcasts` collection must be manually created in the Mongo database. I'm adding it into Thor now, and will need to make sure to add it to Production before we plan to send our first broadcast using this model.

One potential improvement to make admin life easier could be to query the Mobile Commons API to get a list of all broadcasts, and automatically create documents for each broadcast. For now, we'll keep it manual and simple for sake of getting this branch live (after being open for oh, 10 weeks, or so)

aaronschachter committed:
Fixes undefined _id

aaronschachter commented:
Added a test `broadcasts` document with `_id=123` and `campaign=876` and placeholder `messages.signup_prompt_declined`

Tested on Thor:
* [ ] Passing `broadcast_id=123` and `args=YES` sends CampaignBot Menu Signedup message
* [ ] Continuing conversation submits Reportback successfully
* [ ] Passing `broadcast_id=123` and `args=YES` sends the Broadcast Signup Prompt Declined message
* [ ] Texting a Campaign keyword works as expected
* [ ] Posting a `broadcast_id` that doesn't exist in the `broadcasts` collection results in `{"error":{"code":500,"message":"broadcast undefined"}}`
* [ ] Posting a `broadcast_id` that is not associated with an active CampaignBot campaign results in `{"error":{"code":500,"message":"broadcast campaign undefined"}}`

@sergii-tkachenko This PR is ready for review when you have a chance.

aaronschachter commented on app/models/Broadcast.js:
```diff
@@ -0,0 +1,24 @@
+  type: { type: String, enum: ['signup_prompt'] },
+  campaign: Number,
+  messages: {
+    signup_prompt_declined: String,
```
I'm using the nested `messages` convention that was introduced in #763 -- but I'm not entirely convinced this is the way to go right now since we use our Mongo documents as a CMS. Nesting it as an object makes it a little tricker for our admins to manually edit.

sergii-tkachenko replied:
I talked to @jamjensen earlier about that, and he's okay with it. This do add another one level of complexity, but editing documents in Mongo is and advanced technique by itself.

aaronschachter replied:
Ok, will leave it this way. I'm now leaning toward renaming this as `prompt_declined`, to repurpose the property for the future `reportback_prompt` feature we'll need to close out #742

aaronschachter commented on app/models/Broadcast.js:
```diff
@@ -0,0 +1,24 @@
+const broadcastSchema = new mongoose.Schema({
+
+  _id: { type: Number, index: true },
+  type: { type: String, enum: ['signup_prompt'] },
```
Open to other names for this -- eventually there will be another type for `reportback_prompt` per #742

sergii-tkachenko commented on app/models/Broadcast.js:
```diff
@@ -0,0 +1,24 @@
+
+  _id: { type: Number, index: true },
+  type: { type: String, enum: ['signup_prompt'] },
+  campaign: Number,
```
Maybe string? Ids are alphanumeric in Contentful (Phoenix Next): https://www.contentful.com/developers/docs/references/content-delivery-api/#/reference/entries/entry/get-a-single-entry

aaronschachter replied:
Let's leave this as number until we're consuming content from a new API and ready to make the change globally -- it's defined as a `Number` in all of the other models.

sergii-tkachenko commented on app/routes/chatbot.js:
```diff
@@ -112,74 +114,93 @@ router.post('/', (req, res) => {
 
-      return campaign;
-    })
+  return loadCampaign
```
Semicolon.

aaronschachter replied:
Adding a semi-colon where would break the promise chain below. Would it be better if this line read as 
`return loadCampaign.then((campaign) => {`? The other promises are using the same syntax (where the `then` is on a new line).

sergii-tkachenko replied:
Ah, yes, I missed the promise below. Totally my fault.

sergii-tkachenko commented on app/routes/chatbot.js:
```diff
@@ -112,74 +114,93 @@ router.post('/', (req, res) => {
+
+              logger.info(`loaded campaign:${campaign._id}`);
+
+              if (campaign.status === 'closed') {
```
Maybe `!== 'active'` in case we extended campaigns with different states?

aaronschachter replied:
I'd prefer to keep this as is until we cross that bridge. If we had statuses like `featured` and `active`, `!==active` would not work.

sergii-tkachenko commented on app/routes/chatbot.js:
```diff
@@ -112,74 +114,93 @@ router.post('/', (req, res) => {
+          return app.locals.db.broadcasts
+            .findById(scope.broadcast_id)
+            .then((broadcast) => {
+              if (!broadcast) {
```
I'd suggest checking `broadcast._id` instead.

sergii-tkachenko commented on app/routes/chatbot.js:
```diff
@@ -112,74 +114,93 @@ router.post('/', (req, res) => {
+              if (!broadcast) {
+                // TODO: If this happens, store user/broadcast somewhere to recover lost opportunity
+                // to respond.
+                const err = new Error('broadcast undefined');
```
Not sure if `UnprocessibleEntityError` makes sense here?

aaronschachter replied:
Maybe `NotFoundError` ? We can't find a document in the `broadcasts` collection for the `broadcast_id` passed in the request.

sergii-tkachenko replied:
Yes, makes more sense.

sergii-tkachenko commented on app/routes/chatbot.js:
```diff
@@ -112,74 +114,93 @@ router.post('/', (req, res) => {
+            return reject(err);
+          }
+
+          if (campaign.status === 'closed') {
```
Same here.

aaronschachter committed:
Tweaks per #743 review

aaronschachter committed:
Log declined responses to broadcast in bot_requests

sergii-tkachenko commented on app/routes/chatbot.js:
```diff
@@ -112,74 +114,91 @@ router.post('/', (req, res) => {
+        scope.user = user;
+
+        if (scope.broadcast_id) {
+          return app.locals.db.broadcasts
```
Would it make sense to move this to a helper function, something like `loadBroadcastCampaign` to reduce nesting?

aaronschachter replied:
It sounded like it made sense, but realized that we still need the loaded broadcast model [in this line](https://github.com/DoSomething/gambit/blob/f71df68b0b64f19588bd19cd85e660c348cfa2b1/app/routes/chatbot.js#L153) to pass the Broadcast `prompt_declined` message later. Adding a static `loadBroadcast` method to the `broadcast` model would still look the same.

sergii-tkachenko replied:
Let's discuss it tomorrow anyway.

aaronschachter commented on app/routes/chatbot.js:
```diff
@@ -271,8 +290,20 @@ router.post('/', (req, res) => {
+        }
+
+        // Log the no response:
+        app.locals.db.bot_requests.log(req, 'broadcast', null, 'prompt_declined', msg);
```
Seems slightly hacky to pass `broadcast` as `bot_type` and `null` as the `bot_id` -- but we're defining the `prompt_declined` copy on the `broadcast` model. We already are storing the `broadcast_id` within in the `body` object in the model, which is how we currently query the collection to determine YES responses.

<img width="858" alt="screen shot 2017-01-26 at 8 22 04 am" src="https://cloud.githubusercontent.com/assets/1236811/22339946/94b7fdf8-e3a0-11e6-9bb4-8b1e475912e7.png">

aaronschachter commented:
@jamjensen @classicfreddie This was deployed to prod today ([3.3.0](https://github.com/DoSomething/gambit/releases/tag/3.3.0)). I've removed the `CAMPAIGNBOT_BROADCAST_CAMPAIGN` from Heroku and updated the docs: https://github.com/DoSomething/gambit/wiki/Chatbot#signup-broadcasts
